handout/forms/form-builder/form-builder_custom_validation.md

```diff
@@ -0,0 +1,102 @@
+# FormBuilder Custom Validation
+
+As useful as the built-in validators are, it is very useful to be able to include your own.
+Angular 2 allows you to do just that, with minimal effort.
+
+Let's assume we are using the same Login Form, but now we also want to test that our password has an exclamation mark somewhere in it.
+
+_app/login-form.component.ts_
+```ts
+function hasExclamationMark (input: FormControl) {
+  const hasExclamation = input.value.indexOf('!') >= 0;
+
+  return hasExclamation ? null : { needsExclamation: true };
+}
+
+
+// ...
+this.password = new FormControl('', [
+  Validators.required,
+  hasExclamationMark
+]);
+```
+
+I've created a simple function which takes the `FormControl` instance and returns null if everything is fine.
+If the test fails, it returns an object with an arbitrarily named property.
+The property name is what will be used for the `.hasError()` test.
+
+_app/login-form.component.ts_
+```html
+<!-- ... -->
+<div [hidden]="!password.hasError('needsExclamation')">
+  Your password must have an exclamation mark!
+</div>
+<!-- ... -->
+```
+
+### Predefined Parameters
+Having a custom validator check for exclamation marks might be helpful, but what if you need to check for some other form of punctuation?
+It might be overkill to write nearly the same thing over and over again.
+
+Consider the earlier example `Validators.minLength(5)`.
+How did they get away with allowing you to pass in an argument to control the length?
+Simple, really.
+It's not a trick of Angular, or TypeScript.
+It's simple JavaScript closures.
```

Comment:
will everyone reading this book know what a closure is ? i guess a better question is what type of audience are we targeting.